File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/searching/exponential_search.py
try:
    from algorithms.searching.binary_search import binary_search
except ModuleNotFoundError:
    from binary_search import binary_search


def exponential_search(values: list[int], target: int) -> int:
    if not values:
        return -1
    if values[0] == target:
        return 0

    bound = 1
    while bound < len(values) and values[bound] < target:
        bound *= 2

    offset = bound // 2
    local_index = binary_search(values[offset : min(bound + 1, len(values))], target)
    return -1 if local_index == -1 else offset + local_index


if __name__ == "__main__":
    sample = [1, 3, 7, 7, 19, 42, 99]
    target = 19
    print("Exponential Search")
    print(f"input : {sample}")
    print(f"target: {target}")
    print(f"index : {exponential_search(sample, target)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/searching/ExponentialSearch.java
package com.sortingsearching.searching;

import java.util.Arrays;

public final class ExponentialSearch {
    private ExponentialSearch() {
    }

    public static int search(int[] values, int target) {
        if (values.length == 0) {
            return -1;
        }
        if (values[0] == target) {
            return 0;
        }
        int bound = 1;
        while (bound < values.length && values[bound] < target) {
            bound *= 2;
        }
        int offset = bound / 2;
        int[] range = Arrays.copyOfRange(values, offset, Math.min(bound + 1, values.length));
        int localIndex = BinarySearch.search(range, target);
        return localIndex == -1 ? -1 : offset + localIndex;
    }

    public static void main(String[] args) {
        int[] sample = {1, 3, 7, 7, 19, 42, 99};
        int target = 19;
        System.out.println("Exponential Search");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("target: " + target);
        System.out.println("index : " + search(sample, target));
    }
}
